src/Base/Core/Quality/IQuality.hs

```diff
@@ -27,16 +40,19 @@ import Common.Utils (modByFrom)
 3. Minor
 4. Diminished (single, doubly, triply, etc.)
 5. Augmented (single, doubly, triply, etc.)
-
-TODO: Make sure that interval qualities are using smart constructors
-so that diminished and augmented only receive integers >= 1.
 -}
 data Quality
- = Major
- | Perfect
- | Minor
- | Diminished Int
- | Augmented Int
+ = RealMajor
+ | RealPerfect
+ | RealMinor
+ | RealDiminished Int
+ | RealAugmented Int
+
+pattern Major <- RealMajor
+pattern Perfect <- RealPerfect
+pattern Minor <- RealMinor
+pattern Diminished x <- RealDiminished x
+pattern Augmented x <- RealAugmented x
```

Comment:
Here's the magic: I renamed all the actual constructors, and made unidirectional (meaning only usable as patterns) pattern synonyms that have the original names.